flask_session_captcha/__init__.py

```diff
@@ -1,97 +1,159 @@
-            "CAPTCHA_SESSION_KEY", "captcha_answer")
-        self.max = 10**self.digits
-
-        xargs = {}
-        if self.height:
-            xargs['height'] = self.height
-        if self.width:
-            xargs['width'] = self.width
-
-        self.image_generator = ImageCaptcha(**xargs)
-        self.rand = SystemRandom()
-
-        def _generate(css_class=None):
+        self.enabled = app.config.get("CAPTCHA_ENABLE", self.enabled)
+        self.length = app.config.get("CAPTCHA_LENGTH", self.length)
+        self.width = app.config.get("CAPTCHA_WIDTH", self.width)
+        self.height = app.config.get("CAPTCHA_HEIGHT", self.height)
+        self.session_key = app.config.get("CAPTCHA_SESSION_KEY", self.session_key)
+        self.image_generator = ImageCaptcha(**{'width': self.width, 'height': self.height})
+        self.include_alphabet = app.config.get("CAPTCHA_INCLUDE_ALPHABET", self.include_alphabet)
+        self.include_numeric = app.config.get("CAPTCHA_INCLUDE_NUMERIC", self.include_numeric)
+        self.include_punctuation = app.config.get("CAPTCHA_INCLUDE_PUNCTUATION", self.include_punctuation)
+
+
+        def _generate(*args, **kwargs) -> Markup:
+            """Generate Captcha Image"""
             if not self.enabled:
-                return ""
-            base64_captcha = self.generate()
-            data = "data:image/png;base64, {}".format(base64_captcha)
-            if css_class is not None:
-                return Markup("<img src='{}' class='{}'>".format(data, css_class))
-            return Markup("<img src='{}'>".format(data))
+                return Markup(" ")
+
+            base64_captcha = self.generate(*args, **kwargs)
+            data = f"data:image/png;base64, {base64_captcha}"
+            #TODO: replace it with argument base function: users can set ud,class,and ... in template in captcha:filter
+            css = f"class=\'{kwargs.get('css_class')}\'" if kwargs.get('css_class', None) else ''
+            return Markup(f"""<img src='{data}' {css} >""")
 
         app.jinja_env.globals['captcha'] = _generate
 
-        # Check for sessions that do not persist on the server. Issue a warning because
-        # they are most likely open to replay attacks. This addon is built upon flask-session.
-        session_type = app.config.get('SESSION_TYPE')
-        if session_type is None or session_type == "null":
-            raise RuntimeWarning(
-                "Flask-Sessionstore is not set to use a server persistent storage type. This likely means that captchas are vulnerable to replay attacks.")
-        elif session_type == "sqlalchemy":
-            # I have to do this as of version 0.3.1 of flask-session if using sqlalchemy as the session type in order to create the initial database.
-            # Flask-sessionstore seems to have the same problem.
-            app.session_interface.db.create_all()
-
-    def generate(self):
+    def generate(self,*args, **kwargs) -> str:
```

Comment:
This method is called indirectly through templates

```python
 app.jinja_env.globals['captcha'] = _generate
```


_generate method:
```python
def _generate(*args, **kwargs) -> Markup:
  if not self.enabled:
      return Markup(" ")

# actual method call
  base64_captcha = self.generate(*args, **kwargs) #call method with args, kwargs
  
  data = f"data:image/png;base64, {base64_captcha}"
  css = f"class=\'{kwargs.get('css_class')}\'" if kwargs.get('css_class', None) else ''
  return Markup(f"<img src='{data}' {css} >")
````
```python
captcha(hello=True) == _generate(hello=True)
# so 
-generate(hello=True) == generate(hello=True)
```

In the current implementation, the user cannot add an ID or even a dataset or other things to the captcha(only css class). With this implementation method, the user can directly send the captcha arguments in the template and it will be used for the captcha in the main method.(id, dataset, java script event or ...)


```python

{{ captcha(id='id', dataset=" data-test:'true'  ")
or with more argument call
{{ captcha(id='id', dataset=" data-test:'true'; data-ok:'ok' ", numeric=True, alphabet=False, punctuaction=True, onClick='JsFunc(this);')
```

Reply:
with this approach, developer can set a default config for all captchas (config) and for some special captchas developer can directly pass some argument from template filter 
for example i like all my captchas be numeric (across my web site) but for admin login page i like to have a mixed captcha for security purposes